3_train/src/train.py

```diff
@@ -320,7 +329,22 @@ def verbose_print(*args):
         print(f'{epoch}: {datetime.now()} {train_loss}, {valid_loss}', flush=True)
         train_losses.append(train_loss)
         valid_losses.append(valid_loss)
-    return train_losses, valid_losses
+
+        # If this model has the lowest validation loss, save model weights
+        if valid_loss == min(valid_losses):
+            num_epochs_without_improvement = 0
+            saved_epoch = epoch
+            save_weights(model, weights_filepath, overwrite=True)
+        # Otherwise, add to count of high validation loss models and check for early stopping condition
+        else:
+            num_epochs_without_improvement += 1
+            # An early_stopping_patience of -1 means early stopping should not happen
+            if early_stopping_patience != -1:
+                if num_epochs_without_improvement >= early_stopping_patience:
+                    print(f'Stopping at epoch {epoch} after {num_epochs_without_improvement} epochs without improvement', flush=True)
+                    break
```

Comment:
If I'm reading this right, the psuedocode would be:

```
*(only) if lowest validation loss achieved:* # still happens when early stopping is disabled
    *save weights*
if lowest validation loss not achieved:
    track patience
    *if early stopping is enabled:*
        if patience exceeded:
            stop
```
I believe the asterick'd line orders are problematic because if you intended early stopping to be disabled, the code will still be only saving weights from the best epoch (which could be epoch 500 of 1000), thus still giving you early stopped model weights but having done all the non-early stopped epochs.

If you intend to keep the functionality of disabling early stopping, the `if early_stopping_patience != -1` (which indicates if early stopping is disabled/enabled) should be repositioned or duplicated to decide whether we save models weights associated with the best epoch for validation loss or every epoch regardless of better/worse validation loss.

Reply:
Nice catch! You're right. The bug was in my brain - I intentionally coded it this way, not thinking of the possibility that we'd want to save the weights for a model that doesn't have the lowest validation loss. I'll make the change.

Note that weights aren't saved for every epoch - only one set of weights is being saved per training run, for the sake of performance and memory savings. In the future we can add an option to save weights at each epoch if/when we need it.